### PR TITLE
Initial progress

### PR DESCRIPTION
**Pull request by dzenanz**

### CODE REVIEW

**Comment by dzenanz** [comment]
TODO: try https://github.com/Unidata/netcdf-c/pull/2492/ here.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,13 +1,166 @@
...
-  include(ITKModuleExternal)
-else()
-  set(ITK_DIR ${CMAKE_BINARY_DIR})
-  itk_module_impl()
+if(ITK_SOURCE_DIR)
+  message(FATAL_ERROR "${PROJECT_NAME} currently does not support being built as part of ITK_DIR
+  It needs to be built of out source, and pointed to a certain ITK build.")
+endif()
+
+function(dump_cmake_variables)
```
Comment (thewtex) [inline CMakeLists.txt:14]: remove debug code

File: include/itkOMEZarrNGFFImageIO.h [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  using Pointer = SmartPointer<Self>;
+  using ConstPointer = SmartPointer<const Self>;
+
+  /** Method for creation through the object factory. */
+  itkNewMacro(Self);
+
+  /** Run-time type information (and related methods). */
+  itkTypeMacro(OMEZarrNGFFImageIO, ImageIOBase);
+
+  static constexpr unsigned MaximumDimension = 5; // OME-NGFF specifies up to 5D data
```
Comment (thewtex) [inline include/itkOMEZarrNGFFImageIO.h:56]: 5 -> 4 ('c' is the channel / component)

File: src/itkOMEZarrNGFFImageIO.cxx [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+}
+
+
+void
+OMEZarrNGFFImageIO::Write(const void * buffer)
+{
+  int r;
+  netCDF_call(nc_create(getNCFilename(this->m_FileName), NC_CLOBBER, &m_NCID));
+
+  const std::vector<std::string> dimensionNames = { "c", "i", "j", "k", "t" };
```
Comment (thewtex) [inline src/itkOMEZarrNGFFImageIO.cxx:337]: i j k to x y z per OME-Zarr

**Review by thewtex — CHANGES_REQUESTED** [review]
Please get CI passing

**Comment by thewtex** [comment]
> -- Building Shared Libraries:     ON

We should turn off shared libraries if possible.

**Comment by thewtex** [comment]
Linux todo: use ITK's HDF5:

> -- Linking against:      hdf5_cpp-shared;hdf5-shared;/usr/lib/x86_64-linux-gnu/libm.so;itkzlib;/usr/lib/x86_64-linux-gnu/libzip.so;/usr/lib/x86_64-linux-gnu/libbz2.so;libcurl;/usr/lib/x86_64-linux-gnu/libxml2.so

**Comment by thewtex** [comment]
> ZSTD Support:           no

-> yes

https://open.cdash.org/build/8154414/configure

**Comment by thewtex** [comment]
> CMake Error at CMakeLists.txt:55 (add_library):
  add_library cannot create ALIAS target "ZLIB::ZLIB" because target
  "itkzlib" is imported but not globally visible.

**Comment by thewtex** [comment]
For OpenSSL errors, try installing OpenSSL beforehand:

https://github.com/dockcross/dockcross/blob/96d87416f639af0204bdd42553e4b99315ca8476/imagefiles/build-and-install-openssl.sh

Maybe curl on Linux:

https://github.com/dockcross/dockcross/blob/96d87416f639af0204bdd42553e4b99315ca8476/imagefiles/build-and-install-curl.sh

File: CMakeLists.txt [hunk 4]
```diff
@@ -182,7 +182,13 @@ FetchContent_Declare(
 # message("HDF5 after FetchContent_Declare")
 # dump_cmake_variables("HDF5")
 
+set(saved_CXX ${CMAKE_CXX_FLAGS})
+set(saved_C ${CMAKE_C_FLAGS})
+set(CMAKE_CXX_FLAGS -w)
+set(CMAKE_C_FLAGS -w)
 FetchContent_MakeAvailable(netCDF)
+set(CMAKE_CXX_FLAGS ${saved_CXX})
+set(CMAKE_C_FLAGS ${saved_C})
```
Comment (thewtex) [inline CMakeLists.txt:191]: We can add warning exceptions here:

https://github.com/InsightSoftwareConsortium/ITKIOOMEZarrNGFF/blob/44f22e284f020834ca724c9bea9286530dcff7de/.github/workflows/build-test-package.yml#L98-L102

**Comment by dzenanz** [comment]
Closing this PR, as I just pushed version ab1a5a47d0ad625232a29943f0861477f7079f42 to master which works fairly well.

**Pull request closed without merging**